addons/controllers/clusterbootstrap_controller.go

```diff
@@ -278,14 +319,271 @@ func (r *ClusterBootstrapReconciler) createOrPatchclusterBootstrapFromTemplate(c
 	return nil, errors.New("should not happen")
 }
 
-// createOrPatchPackageInstallSecret creates or patches or the secret used for PackageInstall in a cluster
-func (r *ClusterBootstrapReconciler) createOrPatchPackageInstallSecret(cluster *clusterapiv1beta1.Cluster,
-	pkg *runtanzuv1alpha3.ClusterBootstrapPackage, clusterClient client.Client, namespace string, log logr.Logger) (*corev1.Secret, error) {
+// createOrPatchKappPackageInstall contains the logic that create/update PackageInstall CR for kapp-controller on
+// mgmt cluster. The kapp-controller running on mgmt cluster reconciles the PackageInstall CR and creates kapp-controller resources
+// on remote workload cluster. This is required for a workload cluster and its corresponding package installations to be functional.
+func (r *ClusterBootstrapReconciler) createOrPatchKappPackageInstall(clusterBootstrap *runtanzuv1alpha3.ClusterBootstrap, cluster *clusterapiv1beta1.Cluster) error {
+	pkgi := &kapppkgiv1alpha1.PackageInstall{
+		ObjectMeta: metav1.ObjectMeta{
+			// The legacy addons controller uses <cluster name>-<addon name> convention for naming the PackageInstall CR.
+			// https://github.com/vmware-tanzu/tanzu-framework/blob/main/addons/controllers/package_reconciler.go#L195.
+			// util.GeneratePackageInstallName() follows the same pattern.
+			Name: util.GeneratePackageInstallName(cluster.Name, util.GetPackageShortName(clusterBootstrap.Spec.Kapp.RefName)),
+			// kapp-controller PackageInstall CR is installed under the same namespace as tanzuClusterBootstrap. The namespace
+			// is also the same as where the cluster belongs.
+			Namespace: clusterBootstrap.Namespace,
+		},
+	}
+
+	pkgiMutateFn := func() error {
+		// TODO: Followup on the following fields, only add them if needed.
+		// https://github.com/vmware-tanzu/tanzu-framework/issues/1677
+		// if ipkg.ObjectMeta.Annotations == nil {
+		//	 ipkg.ObjectMeta.Annotations = make(map[string]string)
+		// }
+		// ipkg.ObjectMeta.Annotations[addontypes.YttMarkerAnnotation] = ""
+		// ipkg.Spec.SyncPeriod = &metav1.Duration{Duration: r.Config.AppSyncPeriod}
+		// ipkg.Spec.ServiceAccountName = r.Config.AddonServiceAccount
+		pkgi.Spec.PackageRef = &kapppkgiv1alpha1.PackageRef{
+			RefName: clusterBootstrap.Spec.Kapp.RefName,
+			VersionSelection: &versions.VersionSelectionSemver{
+				Prereleases: &versions.VersionSelectionSemverPrereleases{},
+			},
+		}
```

Comment:
Question: Dont see the reference to Cluster and its kubeconfig here. Isnt it needed ?
https://github.com/vmware-tanzu/tanzu-framework/blob/main/addons/controllers/app_reconciler.go#L71

Reply:
yes, for installing kapp-controller resources into remote workload cluster, we do need the reference to Cluster. Good catch.

Reply:
fixed